File: explorer/darwin/ios/lynx_test_bench/TestBenchReplayDataModule.m
// Copyright 2021 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#import "TestBenchReplayDataModule.h"

// "Invoked Method Data" field from record json file
static NSArray *functionCall;
// "Callback" field from record json file
static NSDictionary *callbackData;

static NSArray *jsbIgnoredInfo;

static NSDictionary *jsbSettings;

@implementation TestBenchReplayDataModule

+ (NSString *)name {
  return @"TestBenchReplayDataModule";
}

+ (NSDictionary<NSString *, NSString *> *)methodLookup {
  return @{
    @"getData" : NSStringFromSelector(@selector(getData:)),
  };
}

+ (void)addFunctionCallArray:(NSArray *)responseData {
  functionCall = responseData;
}
+ (void)addCallbackDictionary:(NSDictionary *)callbackDictionary {
  callbackData = callbackDictionary;
}

+ (void)setJSbIgnoredInfo:(NSArray *)info {
  jsbIgnoredInfo = info;
}

+ (void)setJsbSettings:(NSDictionary *)settings {
  jsbSettings = settings;
}

- (void)getData:(LynxCallbackBlock)callback {
  NSMutableDictionary *json = [NSMutableDictionary new];
  [json setObject:[self getRecordData] forKey:@"RecordData"];
  [json setObject:[self getJsbSettings] forKey:@"JsbSettings"];
  [json setObject:[self getJsbIgnoredInfo] forKey:@"JsbIgnoredInfo"];
  NSData *jsonData = [NSJSONSerialization dataWithJSONObject:json options:0 error:0];
  NSString *strData = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
  callback(strData);
}

- (NSString *)getRecordData {
  NSMutableDictionary<NSString *, NSMutableArray *> *json = [NSMutableDictionary new];
  for (int index = 0; index < [functionCall count]; index++) {
    NSDictionary *funcInvoke = functionCall[index];
    NSString *moduleName = [funcInvoke objectForKey:@"Module Name"];
    if (![json objectForKey:moduleName]) {
      [json setObject:[NSMutableArray new] forKey:moduleName];
    }

    NSMutableDictionary *methodLookUp = [NSMutableDictionary new];

    NSString *methodName = [funcInvoke objectForKey:@"Method Name"];

    long requestTime = [[funcInvoke objectForKey:@"Record Time"] doubleValue] * 1000;

    if ([funcInvoke objectForKey:@"RecordMillisecond"]) {
      requestTime = [[funcInvoke objectForKey:@"RecordMillisecond"] doubleValue];
    }

    NSMutableDictionary *params = [funcInvoke objectForKey:@"Params"];

    NSArray *callbackIDs = [params objectForKey:@"callback"];

    NSMutableArray *callbackReturnValues = [NSMutableArray new];

    NSString *functionInvokeLabel = @"";
    for (int i = 0; i < callbackIDs.count; i++) {
      NSDictionary *callbackInfo = [callbackData objectForKey:callbackIDs[i]];
      if (callbackInfo != nil) {
        long responseTime = [[callbackInfo objectForKey:@"Record Time"] doubleValue] * 1000;
        if ([callbackInfo objectForKey:@"RecordMillisecond"]) {
          responseTime = [[callbackInfo objectForKey:@"RecordMillisecond"] doubleValue];
        }
        NSDictionary *callbackKernel = @{
          @"Value" : [callbackInfo objectForKey:@"Params"],
          @"Delay" : @(responseTime - requestTime)
        };
        [callbackReturnValues addObject:callbackKernel];
      }
      functionInvokeLabel = [functionInvokeLabel stringByAppendingFormat:@"%@_", callbackIDs[i]];
    }

    [methodLookUp setObject:methodName forKey:@"Method Name"];
    [methodLookUp setObject:params forKey:@"Params"];
    [methodLookUp setObject:callbackReturnValues forKey:@"Callback"];
    [methodLookUp setObject:functionInvokeLabel forKey:@"Label"];

    if ([funcInvoke objectForKey:@"SyncAttributes"]) {
      [methodLookUp setObject:[funcInvoke objectForKey:@"SyncAttributes"] forKey:@"SyncAttributes"];
    }

    [[json objectForKey:moduleName] addObject:methodLookUp];
  }
  NSData *jsonData = [NSJSONSerialization dataWithJSONObject:json options:0 error:0];
  NSString *strData = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
  return strData;
}

- (NSString *)getJsbIgnoredInfo {
  if (jsbIgnoredInfo != nil) {
    NSData *reData = [NSJSONSerialization dataWithJSONObject:jsbIgnoredInfo options:0 error:0];
    NSString *infoData = [[NSString alloc] initWithData:reData encoding:NSUTF8StringEncoding];
    return infoData;
  } else {
    return @"[]";
  }
}

- (NSString *)getJsbSettings {
  if (jsbSettings != nil) {
    NSData *reData = [NSJSONSerialization dataWithJSONObject:jsbSettings options:0 error:0];
    NSString *settingData = [[NSString alloc] initWithData:reData encoding:NSUTF8StringEncoding];
    return settingData;
  } else {
    return @"{}";
  }
}

@end

File: explorer/android/lynx_test_bench/src/main/java/com/lynx/testbench/TestBenchReplayDataModule.java
// Copyright 2021 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.testbench;

import android.content.Context;
import com.lynx.jsbridge.LynxMethod;
import com.lynx.jsbridge.LynxModule;
import com.lynx.react.bridge.Callback;
import com.lynx.tasm.base.LLog;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class TestBenchReplayDataModule extends LynxModule {
  private static final String TAG = "TestBenchReplayDataModule";
  // "Invoked Method Data" field from record json file
  private static JSONArray mFunctionCall;
  // "Callback" field from record json file
  private static JSONObject mCallbackData;
  // replay additional info
  private static JSONArray mJsbIgnoredInfo;
  // "jsbSettings" field from record json file
  private static JSONObject mJsbSettings;

  public TestBenchReplayDataModule(Context context) {
    super(context);
  }

  public static void addFunctionCallArray(JSONArray responseData) {
    mFunctionCall = responseData;
  }

  public static void addCallbackDictionary(JSONObject callbackDictionary) {
    mCallbackData = callbackDictionary;
  }

  public static void setJsbIgnoredInfo(JSONArray jsbIgnoredInfo) {
    mJsbIgnoredInfo = jsbIgnoredInfo;
  }

  public static void setJsbSettings(JSONObject jsbSettings) {
    mJsbSettings = jsbSettings;
  }

  @LynxMethod
  public void getData(final Callback callback) {
    JSONObject data = new JSONObject();
    try {
      data.put("RecordData", getRecordData());
      data.put("JsbIgnoredInfo", getJsbIgnoredInfo());
      data.put("JsbSettings", getJsbSettings());
    } catch (JSONException e) {
      LLog.e("TestBench", "Record file format error!");
      callback.invoke("{}");
    }
    callback.invoke(data.toString());
  }

  private String getRecordData() {
    JSONObject json = new JSONObject();
    try {
      for (int index = 0; index < mFunctionCall.length(); index++) {
        JSONObject funcInvoke = mFunctionCall.getJSONObject(index);
        String moduleName = funcInvoke.getString("Module Name");
        if (!json.has(moduleName)) {
          json.put(moduleName, new JSONArray());
        }

        JSONObject methodLookUp = new JSONObject();

        String methodName = funcInvoke.getString("Method Name");

        long requestTime = Long.parseLong(funcInvoke.getString("Record Time")) * 1000;
        if (funcInvoke.has("RecordMillisecond")) {
          requestTime = funcInvoke.getLong("RecordMillisecond");
        }
        JSONObject params = funcInvoke.getJSONObject("Params");

        JSONArray callbackIDs;
        try {
          callbackIDs = params.getJSONArray("callback");
        } catch (JSONException e) {
          callbackIDs = null;
        }
        StringBuffer functionInvokeLabel = new StringBuffer();
        JSONArray callbackReturnValues = new JSONArray();
        if (callbackIDs != null) {
          for (int i = 0; i < callbackIDs.length(); i++) {
            JSONObject callbackInfo;
            try {
              callbackInfo = mCallbackData.getJSONObject(callbackIDs.getString(i));
            } catch (JSONException e) {
              callbackInfo = null;
            }
            if (callbackInfo != null) {
              // get callback response Time
              long responseTime = Long.parseLong(callbackInfo.getString("Record Time")) * 1000;
              if (funcInvoke.has("RecordMillisecond")) {
                responseTime = callbackInfo.getLong("RecordMillisecond");
              }
              JSONObject callbackKernel = new JSONObject();
              callbackKernel.put("Value", callbackInfo.getJSONObject("Params"));
              callbackKernel.put("Delay", responseTime - requestTime);
              callbackReturnValues.put(i, callbackKernel);
            }
            functionInvokeLabel.append(callbackIDs.getString(i)).append("_");
          }
        }

        if (funcInvoke.has("SyncAttributes")) {
          methodLookUp.put("SyncAttributes", funcInvoke.getJSONObject("SyncAttributes"));
        }

        methodLookUp.put("Method Name", methodName);
        methodLookUp.put("Params", params);
        methodLookUp.put("Callback", callbackReturnValues);
        methodLookUp.put("Label", functionInvokeLabel.toString());

        json.getJSONArray(moduleName).put(methodLookUp);
      }
    } catch (JSONException e) {
      LLog.e("TestBench", "Record file format error!");
      return "{}";
    }
    return json.toString();
  }

  private String getJsbIgnoredInfo() {
    if (mJsbIgnoredInfo != null) {
      return mJsbIgnoredInfo.toString();
    } else {
      LLog.e("TestBench",
          "getJsbIgnoredInfo "
              + " error: download File failed");
      return "[]";
    }
  }

  private String getJsbSettings() {
    if (mJsbSettings != null) {
      return mJsbSettings.toString();
    } else {
      LLog.e("TestBench",
          "getJsbSettings "
              + " error: download File failed");
      return "{}";
    }
  }
}
